### PR TITLE
Refactor for Extensibility

### PR DESCRIPTION
**Pull request by davilima6**

### CODE REVIEW

**Comment by davilima6** [comment]
I'm not sure if we should include the last commit, which is a test case.

When testing the case for mapping widgets by ID, I got the following error:

```
 FAIL  src/components/manage/Form/Field.test.jsx
  ● Console

    console.error node_modules/react-test-renderer/cjs/react-test-renderer.development.js:5120
      The above error occurred in the <InjectIntl(Connect(SchemaWidget))> component:
          in InjectIntl(Connect(SchemaWidget)) (created by DragDropContext(InjectIntl(Connect(SchemaWidget))))
          in DragDropContext(InjectIntl(Connect(SchemaWidget))) (created by Field)
          in Field

      Consider adding an error boundary to your tree to customize error handling behavior.
      Visit https://fb.me/react-error-boundaries to learn more about error boundaries.

  ● Field › renders a field with a specific id

    Invariant Violation: [React Intl] Could not find required `intl` object. <IntlProvider> needs to exist in the component ancestry.

      41 |       },
      42 |     });
    > 43 |     const component = renderer.create(<Field id="schema" />);
      44 |     const json = component.toJSON();
      45 |     expect(json).toMatchSnapshot();
      46 |   });

      at invariant (node_modules/react-intl/node_modules/invariant/invariant.js:42:15)
      at invariantIntlContext (node_modules/react-intl/lib/index.js:358:3)
      at new InjectIntl (node_modules/react-intl/lib/index.js:431:7)
      at constructClassInstance (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:2310:20)
      at updateClassComponent (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:3817:9)
      at beginWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:4459:16)
      at performUnitOfWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7271:16)
      at workLoop (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7294:26)
      at renderRoot (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7325:9)
      at performWorkOnRoot (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7881:24)
      at performWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7802:9)
      at performSyncWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7779:5)
      at requestWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7679:7)
      at scheduleWorkImpl (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7554:13)
      at scheduleWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7514:12)
      at scheduleRootUpdate (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8142:5)
      at updateContainerAtExpirationTime (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8170:12)
      at Object.updateContainer (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8189:14)
      at Object.create (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8866:18)
      at Object.<anonymous> (src/components/manage/Form/Field.test.jsx:43:32)
```

To overcome that I wrapped <Field/> with `injectIntl` and thus had to mock store for the other test cases. Now it fails with:

```
yarn test:unit                                                                                                                             17:02:58
yarn run v1.7.0
$ better-npm-run test_unit
running better-npm-run in /Users/d.medeiros/Projects/_plone/plone-react
Executing script: test_unit

to be executed: jest --no-cache --runInBand
 PASS  src/reducers/users/users.test.js
 FAIL  src/components/manage/Form/Field.test.jsx (8.895s)
  ● Console

    console.error node_modules/react-test-renderer/cjs/react-test-renderer.development.js:5120
      The above error occurred in the <Connect(SchemaWidget)> component:
          in Connect(SchemaWidget) (created by InjectIntl(Connect(SchemaWidget)))
          in InjectIntl(Connect(SchemaWidget)) (created by DragDropContext(InjectIntl(Connect(SchemaWidget))))
          in DragDropContext(InjectIntl(Connect(SchemaWidget))) (created by Field)
          in Field (created by InjectIntl(Field))
          in InjectIntl(Field)
          in IntlProvider (created by Connect(IntlProvider))
          in Connect(IntlProvider) (created by Provider)
          in Provider (created by Provider)
          in Provider

      Consider adding an error boundary to your tree to customize error handling behavior.
      Visit https://fb.me/react-error-boundaries to learn more about error boundaries.

  ● Field › renders a field with a specific id

    SyntaxError: Unexpected token u in JSON at position 0
        at JSON.parse (<anonymous>)

       99 | @injectIntl
      100 | @connect((state, props) => ({
    > 101 |   value: JSON.parse(props.value),
      102 | }))
      103 | /**
      104 |  * SchemaWidget component class.

      at Function.mapToProps (src/components/manage/Widgets/SchemaWidget.jsx:101:15)
      at mapToPropsProxy (node_modules/react-redux/lib/connect/wrapMapToProps.js:54:46)
      at Function.detectFactoryAndVerify (node_modules/react-redux/lib/connect/wrapMapToProps.js:63:19)
      at mapToPropsProxy (node_modules/react-redux/lib/connect/wrapMapToProps.js:54:46)
      at handleFirstCall (node_modules/react-redux/lib/connect/selectorFactory.js:37:18)
      at pureFinalPropsSelector (node_modules/react-redux/lib/connect/selectorFactory.js:85:81)
      at Object.runComponentSelector [as run] (node_modules/react-redux/lib/components/connectAdvanced.js:43:25)
      at Connect.initSelector (node_modules/react-redux/lib/components/connectAdvanced.js:195:23)
      at new Connect (node_modules/react-redux/lib/components/connectAdvanced.js:136:15)
      at constructClassInstance (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:2310:20)
      at updateClassComponent (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:3817:9)
      at beginWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:4459:16)
      at performUnitOfWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7271:16)
      at workLoop (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7294:26)
      at renderRoot (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7325:9)
      at performWorkOnRoot (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7881:24)
      at performWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7802:9)
      at performSyncWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7779:5)
      at requestWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7679:7)
      at scheduleWorkImpl (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7554:13)
      at scheduleWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7514:12)
      at scheduleRootUpdate (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8142:5)
      at updateContainerAtExpirationTime (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8170:12)
      at Object.updateContainer (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8189:14)
      at Object.create (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:8866:18)
      at Object.<anonymous> (src/components/manage/Form/Field.test.jsx:43:32)
```

File: src/components/manage/Form/Field.jsx [hunk 1]
```diff
@@ -22,7 +22,11 @@ const getWidgetByFieldId = id => widgetMapping.id[id] || null;
  * @param {string} widget Widget
  * @returns {string} Widget component.
  */
-const getWidgetByName = widget => widgetMapping.widget[widget] || null;
+const getWidgetByName = widget => {
+  return typeof widget === 'string'
+    ? widgetMapping.widget[widget] || getWidgetDefault()
+    : null;
+};
```
Comment (davilima6) [inline src/components/manage/Form/Field.jsx]: @sneridagh: implementing https://github.com/plone/plone-react/blob/bccdb930f64a3025452de66413ddc33715ac49e4/src/components/manage/Form/Field.jsx#L47-L49

**Comment by sneridagh** [comment]
@davilima6 all good now! :)

File: src/components/theme/View/View.jsx [hunk 2]
```diff
@@ -202,7 +202,9 @@ export default class View extends Component {
       <div id="view">
         <Helmet
           bodyAttributes={{
-            class: `view-${RenderedView.displayName.toLowerCase()}`,
+            class: RenderedView.displayName
+              ? `view-${RenderedView.displayName.toLowerCase()}`
+              : 'view-undefined',
```
Comment (davilima6) [inline src/components/theme/View/View.jsx]: @sneridagh, in which cases it gets undefined? Doesn't the new plugin always work?

**Comment by sneridagh** [comment]
@robgietema @davilima6 I think I've completed all that I wanted to accomplish with this first batch of refactor for ease extensibility. Please review it and tell me what do you think. I think having this on place is a must in the short term for projects and GSOC. I would merge it during this week sprint.

/cc @tisto

File: src/components/theme/View/View.jsx [hunk 3]
```diff
@@ -215,10 +202,14 @@ export default class View extends Component {
       <div id="view">
         <Helmet
           bodyAttributes={{
-            class: `view-${viewName.toLowerCase()}`,
+            class: RenderedView.displayName
+              ? `view-${RenderedView.displayName.toLowerCase()}`
+              : 'view-undefined',
```
Comment (davilima6) [inline src/components/theme/View/View.jsx]: maybe `view-default` or something else would look better? `undefined` sounds a bit too javascripty to me, like an error

File: .babelrc [hunk 4]
```diff
@@ -5,7 +5,10 @@
     "add-module-exports",
     "transform-decorators-legacy",
     "transform-class-properties",
-    "transform-react-display-name"
```
Comment (davilima6) [inline .babelrc]: A note: I think dropped dependencies are not removed from yarn.lock automatically. But we can worry about that when making a release

**Comment by davilima6** [comment]
LVGTM :) thanks @sneridagh !

**Comment by sneridagh** [comment]
@davilima6 Merged!! :)

**Comment by davilima6** [comment]
Nice! Wish I was there, enjoy! :)

**Pull request merged**